src/shogun/lib/parameter_observers/ParameterObserverLogger.cpp

```diff
@@ -0,0 +1,33 @@
+/*
+* Written (W) 2019 Giovanni De Toni
+*/
+
+#include <shogun/lib/parameter_observers/ParameterObserverLogger.h>
+//#include <shogun/lib/type_case.h>
```

Comment:
As a side not, I am not able to include this header file at all (that's why everything down is commented out). I keep getting this strange linker error here:
```
/usr/bin/ld.gold: error: CMakeFiles/libshogun.dir/lib/parameter_observers/ParameterObserverLogger.cpp.o: multiple definition of 'shogun::type_internal::print_map[abi:cxx11](std::unordered_map<std::type_index, shogun::TYPE, std::hash<std::type_index>, std::equal_to<std::type_index>, std::allocator<std::pair<std::type_index const, shogun::TYPE> > > const&)'
/usr/bin/ld.gold: CMakeFiles/libshogun.dir/io/TBOutputFormat.cpp.o: previous definition here
/usr/bin/ld.gold: error: CMakeFiles/libshogun.dir/lib/parameter_observers/ParameterObserverLogger.cpp.o: multiple definition of 'shogun::type_internal::get_type(shogun::Any const&, std::unordered_map<std::type_index, shogun::TYPE, std::hash<std::type_index>, std::equal_to<std::type_index>, std::allocator<std::pair<std::type_index const, shogun::TYPE> > > const&)'
/usr/bin/ld.gold: CMakeFiles/libshogun.dir/io/TBOutputFormat.cpp.o: previous definition here
collect2: error: ld returned 1 exit status
```
Maybe the solution is straightforward but I can't see it :(

Reply:
@gf712 any ideas?

Reply:
Hmmm ok, there are two ways to solve this I think. Either move the definition of both those function into a translation unit (keep declaration where it is though), or mark them as `inline` in `type_case.h` (but inline is not enforceable though, so could break in some compilers/compiler versions). I can do that in a separate PR? It would be just good to check if that is the issue.

Reply:
Actually `inline` is fine here because the linker will always see it as inline (and it won't be an ODR violation), whether or not the compiler decides to inline it in the machine code.

Reply:
Okay then, I may try to fix it here in this PR.

Reply:
Issue solved.